backend/src/main/scala/cromwell/backend/wdl/OutputEvaluator.scala

```diff
@@ -1,71 +1,80 @@
 package cromwell.backend.wdl
 
-import cats.Applicative
 import cats.data.EitherT._
 import cats.data.Validated.{Invalid, Valid}
 import cats.data.{EitherT, NonEmptyList}
-import cats.instances.list._
 import cats.instances.try_._
+import cats.syntax.apply._
 import cats.syntax.either._
-import cats.syntax.traverse._
+import cats.syntax.validated._
 import cromwell.backend.BackendJobDescriptor
 import cromwell.core.{CallOutputs, JobOutput}
 import lenthall.validation.ErrorOr.ErrorOr
 import lenthall.validation.Validation._
 import wdl.types.WdlType
 import wdl.values.WdlValue
 import wom.callable.Callable.OutputDefinition
-import wom.expression.{IoFunctionSet, WomExpression}
+import wom.expression.IoFunctionSet
 
 import scala.util.{Failure, Success, Try}
 object OutputEvaluator {
   sealed trait EvaluatedJobOutputs
   case class ValidJobOutputs(outputs: CallOutputs) extends EvaluatedJobOutputs
   case class InvalidJobOutputs(errors: NonEmptyList[String]) extends EvaluatedJobOutputs
   case class JobOutputsEvaluationException(exception: Throwable) extends EvaluatedJobOutputs
-  
+
   type OutputResult[A] = EitherT[Try, NonEmptyList[String], A]
 
   def evaluateOutputs(jobDescriptor: BackendJobDescriptor,
                       ioFunctions: IoFunctionSet,
                       postMapper: WdlValue => Try[WdlValue] = v => Success(v)): EvaluatedJobOutputs = {
-    val knownValues = jobDescriptor.inputDeclarations map {
-      case (declaration, value) => declaration.name -> value
-    }
+    val knownValues = jobDescriptor.localInputs
     
-    def evaluateOutputExpression(expression: WomExpression): OutputResult[WdlValue] = {
-      EitherT { Try(expression.evaluateValue(knownValues, ioFunctions)).map(_.toEither) }
+    def foldFunction(accumulatedOutputs: Try[ErrorOr[List[(String, WdlValue)]]], output: OutputDefinition) = accumulatedOutputs flatMap { accumulated =>
+      // Extract the valid pairs from the job outputs accumulated so far, and add to it the inputs (outputs can also reference inputs)
+      val allKnownValues = accumulated match {
+        case Valid(outputs) => outputs.toMap[String, WdlValue] ++ knownValues
+        case Invalid(_) => knownValues
```

Comment:
This logic seems fishy.  Seems like knownValues should be passed into initial state and then invalid is really invalid no?

Reply:
If we pass `knownValues` (which contains input values) in the initial state then it will also end up in the final map, and we only want the outputs.